Fix: Avoid infinite recursive calls

kzkn opened the pull request:
When I try to use IME (e.g. fcitx) to narrow down a combo box (e.g. Description, Transfer in Account), infinite recursion occurs via `popup_set_focus` and GnuCash falls down with SIGSEGV.

jralls commented:
Please post a stack trace of the recursion. The only two implementations of `popup_set_focus` just call ` gtk_widget_grab_focus` so I'm puzzled how that would result in another call of `gnc_item_edit_show_popup`, never mind how an IME would make a difference.

jralls commented on gnucash/register/register-gnome/gnucash-item-edit.c:
```diff
@@ -1112,9 +1112,12 @@ gnc_item_edit_show_popup (GncItemEdit *item_edit)
     item_edit->show_popup = TRUE;
 
-    if (item_edit->popup_set_focus)
+    if (item_edit->popup_set_focus && !item_edit->set_focus) {
```
Curly braces get a line of their own, see for example a few lines down.

kzkn commented:
@jralls 
Thanks for the comment.

> Please post a stack trace of the recursion.

Here is. This is the output from running the binary that built [current HEAD of stable branch](919bfa717833e8a5775d5214af09908a157f81b1).
```
#0  0x00007ffaedf3094e in pcre2_compile_8
    (pattern=pattern@entry=0x55d25ffcde90 ":", patlen=patlen@entry=18446744073709551615, options=options@entry=1074397184, errorptr=errorptr@entry=0x7ffc352ff4ec, erroroffset=erroroffset@entry=0x7ffc352ff4f0, ccontext=ccontext@entry=0x55d2606962f0) at src/pcre2_compile.c:9766
#1  0x00007ffaf6727054 in regex_compile (pattern=pattern@entry=0x55d25ffcde90 ":", compile_options=1074397184, 
    compile_options@entry=524288, newline_options=4, bsr_options=1, error=error@entry=0x0) at ../glib/glib/gregex.c:1814
#2  0x00007ffaf67278ff in g_regex_new (pattern=0x55d25ffcde90 ":", compile_options=G_REGEX_DEFAULT, match_options=G_REGEX_MATCH_DEFAULT, error=0x0)
    at ../glib/glib/gregex.c:1751
#3  0x00007ffaf8d61ff7 in gnc_combo_cell_type_ahead_search (newval=0x55d26069a450 "おやつ", full_store=0x55d25fec45b0, cell=0x55d25feac730)
    at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-gnome/combocell-gnome.c:603
#4  0x00007ffaf8d62517 in gnc_combo_cell_modify_verify
    (_cell=0x55d25feac730, change=0x55d26069a430 "おやつ", change_len=9, newval=0x55d26069a450 "おやつ", newval_len=9, cursor_position=0x7ffc352ffd54, start_selection=0x7ffc352ff80c, end_selection=0x7ffc352ff810) at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-gnome/combocell-gnome.c:730
#5  0x00007ffaf8d493fe in gnc_table_modify_update
    (table=0x55d25feb1600, virt_loc=..., change=0x55d26069a430 "おやつ", change_len=9, newval=0x55d26069a450 "おやつ", newval_len=9, cursor_position=0x7ffc352ffd54, start_selection=0x7ffc352ff80c, end_selection=0x7ffc352ff810, cancelled=0x55d25ff1430c)
    at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-core/table-allgui.c:1269
#6  0x00007ffaf8d733f6 in gnucash_sheet_insert_cb
    (editable=0x55d25ff150b0, insert_text=0x55d26069a430 "おやつ", insert_text_len=9, position=0x7ffc352ffd54, sheet=0x55d25ff14260)
    at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-gnome/gnucash-sheet.c:1076
#7  0x00007ffaf7d5e146 in g_closure_invoke
    (closure=0x55d25f7fc0b0, return_value=0x0, n_param_values=4, param_values=0x7ffc352ffa00, invocation_hint=0x7ffc352ff980) at ../glib/gobject/gclosure.c:832
#8  0x00007ffaf7d8c2f8 in signal_emit_unlocked_R.isra.0
    (node=node@entry=0x55d25f707680, detail=detail@entry=0, instance=instance@entry=0x55d25ff150b0, emission_return=emission_return@entry=0x0, instance_and_params=instance_and_params@entry=0x7ffc352ffa00) at ../glib/gobject/gsignal.c:3802
#9  0x00007ffaf7d7c095 in g_signal_emit_valist
    (instance=instance@entry=0x55d25ff150b0, signal_id=signal_id@entry=311, detail=<optimized out>, var_args=var_args@entry=0x7ffc352ffc00)
    at ../glib/gobject/gsignal.c:3555
#10 0x00007ffaf7d7c544 in g_signal_emit_by_name (instance=instance@entry=0x55d25ff150b0, detailed_signal=detailed_signal@entry=0x7ffaf86295c7 "insert-text")
    at ../glib/gobject/gsignal.c:3654
#11 0x00007ffaf834969b in gtk_entry_insert_text (editable=0x55d25ff150b0, new_text=0x55d26069a410 "おやつ", new_text_length=9, position=0x7ffc352ffd54)
    at ../gtk/gtk/gtkentry.c:5175
#12 0x00007ffaf8350e28 in gtk_entry_enter_text (entry=0x55d25ff150b0, str=0x55d26069a410 "おやつ") at ../gtk/gtk/gtkentry.c:6172
#13 0x00007ffaf7d5cb10 in g_cclosure_marshal_VOID__STRINGv
    (closure=0x55d25ff153a0, return_value=<optimized out>, instance=0x55d25ff14dc0, args=<optimized out>, marshal_data=<optimized out>, n_params=<optimized out>, param_types=0x55d25f717a80) at ../glib/gobject/gmarshal.c:1462
#14 0x00007ffaf7d7b451 in _g_closure_invoke_va
    (param_types=0x55d25f717a80, n_params=1, args=0x7ffc352fff90, instance=0x55d25ff14dc0, return_value=0x0, closure=0x55d25ff153a0)
    at ../glib/gobject/gclosure.c:895
#15 g_signal_emit_valist (instance=instance@entry=0x55d25ff14dc0, signal_id=signal_id@entry=331, detail=<optimized out>, var_args=var_args@entry=0x7ffc352fff90)
    at ../glib/gobject/gsignal.c:3462
#16 0x00007ffaf7d7c544 in g_signal_emit_by_name (instance=0x55d25ff14dc0, detailed_signal=0x7ffaf862a211 "commit") at ../glib/gobject/gsignal.c:3654
#17 0x00007ffaf7d5cb10 in g_cclosure_marshal_VOID__STRINGv
    (closure=0x55d25ffbfad0, return_value=<optimized out>, instance=0x55d25ff85cd0, args=<optimized out>, marshal_data=<optimized out>, n_params=<optimized out>, param_types=0x55d25f717a80) at ../glib/gobject/gmarshal.c:1462
#18 0x00007ffaf7d7b451 in _g_closure_invoke_va
    (param_types=0x55d25f717a80, n_params=1, args=0x7ffc353002a0, instance=0x55d25ff85cd0, return_value=0x0, closure=0x55d25ffbfad0)
    at ../glib/gobject/gclosure.c:895
#19 g_signal_emit_valist (instance=0x55d25ff85cd0, signal_id=331, detail=<optimized out>, var_args=var_args@entry=0x7ffc353002a0) at ../glib/gobject/gsignal.c:3462
#20 0x00007ffaf7d7c324 in g_signal_emit (instance=<optimized out>, signal_id=<optimized out>, detail=<optimized out>) at ../glib/gobject/gsignal.c:3612
#21 0x00007ffac2f8b58d in fcitx_im_context_commit_string (str=<optimized out>, context=0x55d25ff85cd0)
    at /usr/src/debug/fcitx5-gtk/fcitx5-gtk-5.0.23/gtk3/fcitximcontext.cpp:876
#22 fcitx_im_context_commit_preedit(FcitxIMContext*) (context=0x55d25ff85cd0) at /usr/src/debug/fcitx5-gtk/fcitx5-gtk-5.0.23/gtk3/fcitximcontext.cpp:890
#23 0x00007ffac2f8b6ab in fcitx_im_context_commit_preedit (context=0x55d25ff85cd0) at /usr/src/debug/fcitx5-gtk/fcitx5-gtk-5.0.23/gtk3/fcitximcontext.cpp:885
#24 fcitx_im_context_focus_out (context=0x55d25ff85cd0) at /usr/src/debug/fcitx5-gtk/fcitx5-gtk-5.0.23/gtk3/fcitximcontext.cpp:908
#25 fcitx_im_context_focus_out(GtkIMContext*) (context=0x55d25ff85cd0) at /usr/src/debug/fcitx5-gtk/fcitx5-gtk-5.0.23/gtk3/fcitximcontext.cpp:897
#26 0x00007ffaf83527f9 in gtk_entry_focus_out (widget=0x55d25ff150b0, event=event@entry=0x7ffc35300710) at ../gtk/gtk/gtkentry.c:5015
#27 0x00007ffaf8287c89 in _gtk_marshal_BOOLEAN__BOXEDv
    (closure=0x55d25ee8c700, return_value=0x7ffc35300510, instance=<optimized out>, args=<optimized out>, marshal_data=<optimized out>, n_params=<optimized out>, param_types=0x55d25ef1e620) at gtk/gtkmarshalers.c:130
#28 0x00007ffaf7d7b451 in _g_closure_invoke_va
    (param_types=0x55d25ef1e620, n_params=1, args=0x7ffc353005c0, instance=0x55d25ff150b0, return_value=0x7ffc35300510, closure=0x55d25ee8c700)
    at ../glib/gobject/gclosure.c:895
#29 g_signal_emit_valist (instance=0x55d25ff150b0, signal_id=77, detail=<optimized out>, var_args=var_args@entry=0x7ffc353005c0) at ../glib/gobject/gsignal.c:3462
#30 0x00007ffaf7d7c324 in g_signal_emit (instance=instance@entry=0x55d25ff150b0, signal_id=<optimized out>, detail=detail@entry=0)
    at ../glib/gobject/gsignal.c:3612
#31 0x00007ffaf8544b25 in gtk_widget_event_internal.part.0.lto_priv.0 (widget=0x55d25ff150b0, event=0x7ffc35300710) at ../gtk/gtk/gtkwidget.c:7812
#32 0x00007ffaf8d68712 in gnc_item_edit_focus_out (item_edit=0x55d25ff14990)
    at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-gnome/gnucash-item-edit.c:312
#33 0x00007ffaf8d73edc in gnucash_sheet_focus_out_event (widget=0x55d25ff14260, event=0x55d26069a340)
    at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-gnome/gnucash-sheet.c:1275
#34 0x00007ffaf8284fd8 in _gtk_marshal_BOOLEAN__BOXED
    (closure=0x55d25ee8c700, return_value=0x7ffc353008d0, n_param_values=<optimized out>, param_values=0x7ffc35300930, invocation_hint=<optimized out>, marshal_data=<optimized out>) at gtk/gtkmarshalers.c:84
#35 0x00007ffaf7d5e146 in g_closure_invoke
    (closure=0x55d25ee8c700, return_value=0x7ffc353008d0, n_param_values=2, param_values=0x7ffc35300930, invocation_hint=0x7ffc353008b0)
    at ../glib/gobject/gclosure.c:832
#36 0x00007ffaf7d8c427 in signal_emit_unlocked_R.isra.0
    (node=<optimized out>, detail=detail@entry=0, instance=instance@entry=0x55d25ff14260, emission_return=emission_return@entry=0x7ffc35300a40, instance_and_params=instance_and_params@entry=0x7ffc35300930) at ../glib/gobject/gsignal.c:3841
#37 0x00007ffaf7d7baa0 in g_signal_emit_valist
    (instance=<optimized out>, signal_id=<optimized out>, detail=<optimized out>, var_args=var_args@entry=0x7ffc35300af0) at ../glib/gobject/gsignal.c:3565
#38 0x00007ffaf7d7c324 in g_signal_emit (instance=instance@entry=0x55d25ff14260, signal_id=<optimized out>, detail=detail@entry=0)
    at ../glib/gobject/gsignal.c:3612
#39 0x00007ffaf8544b25 in gtk_widget_event_internal.part.0.lto_priv.0 (widget=0x55d25ff14260, event=0x55d26069a340) at ../gtk/gtk/gtkwidget.c:7812
#40 0x00007ffaf8543d8b in gtk_widget_send_focus_change (widget=0x55d25ff14260, event=0x55d26069a340) at ../gtk/gtk/gtkwidget.c:16244
#41 0x00007ffaf8552b52 in do_focus_change (widget=0x55d25ff14260, in=0) at ../gtk/gtk/gtkwindow.c:8487
#42 0x00007ffaf855320a in gtk_window_real_set_focus (window=<optimized out>, focus=0x55d25ffca9d0) at ../gtk/gtk/gtkwindow.c:8776
#43 0x00007ffaf7d5cca6 in g_cclosure_marshal_VOID__OBJECTv
    (closure=0x55d25ee6c510, return_value=<optimized out>, instance=0x55d25efe87b0, args=<optimized out>, marshal_data=<optimized out>, n_params=<optimized out>, param_types=0x55d25ef0bb20) at ../glib/gobject/gmarshal.c:1910
#44 0x00007ffaf7d7b451 in _g_closure_invoke_va
    (param_types=0x55d25ef0bb20, n_params=1, args=0x7ffc35300ea0, instance=0x55d25efe87b0, return_value=0x0, closure=0x55d25ee6c510)
    at ../glib/gobject/gclosure.c:895
#45 g_signal_emit_valist (instance=0x55d25efe87b0, signal_id=114, detail=<optimized out>, var_args=var_args@entry=0x7ffc35300ea0) at ../glib/gobject/gsignal.c:3462
#46 0x00007ffaf7d7c324 in g_signal_emit (instance=<optimized out>, signal_id=<optimized out>, detail=<optimized out>) at ../glib/gobject/gsignal.c:3612
#47 0x00007ffaf8517e6a in gtk_tree_view_grab_focus (widget=0x55d25ffca9d0) at ../gtk/gtk/gtktreeview.c:8823
#48 0x00007ffaf7d7b451 in _g_closure_invoke_va
    (param_types=0x0, n_params=0, args=0x7ffc35301150, instance=0x55d25ffca9d0, return_value=0x0, closure=0x55d25ee551a0) at ../glib/gobject/gclosure.c:895
#49 g_signal_emit_valist (instance=0x55d25ffca9d0, signal_id=57, detail=<optimized out>, var_args=var_args@entry=0x7ffc35301150) at ../glib/gobject/gsignal.c:3462
#50 0x00007ffaf7d7c324 in g_signal_emit (instance=instance@entry=0x55d25ffca9d0, signal_id=<optimized out>, detail=detail@entry=0)
    at ../glib/gobject/gsignal.c:3612
#51 0x00007ffaf853247c in gtk_widget_grab_focus (widget=0x55d25ffca9d0) at ../gtk/gtk/gtkwidget.c:8101
#52 0x00007ffaf8d63028 in popup_set_focus (widget=0x55d25ffc24c0, user_data=0x55d25fd89960)
    at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-gnome/combocell-gnome.c:1052
#53 0x00007ffaf8d6aab9 in gnc_item_edit_show_popup (item_edit=0x55d25ff14990)
    at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-gnome/gnucash-item-edit.c:1116
#54 0x00007ffaf8d622b3 in gnc_combo_cell_type_ahead_search (newval=0x55d260699900 "おやつ", full_store=0x55d25fec45b0, cell=0x55d25feac730)
    at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-gnome/combocell-gnome.c:652
#55 0x00007ffaf8d62517 in gnc_combo_cell_modify_verify
    (_cell=0x55d25feac730, change=0x55d2606998e0 "おやつ", change_len=9, newval=0x55d260699900 "おやつ", newval_len=9, cursor_position=0x7ffc35301af4, start_selection=0x7ffc353015ac, end_selection=0x7ffc353015b0) at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-gnome/combocell-gnome.c:730
#56 0x00007ffaf8d493fe in gnc_table_modify_update
    (table=0x55d25feb1600, virt_loc=..., change=0x55d2606998e0 "おやつ", change_len=9, newval=0x55d260699900 "おやつ", newval_len=9, cursor_position=0x7ffc35301af4, start_selection=0x7ffc353015ac, end_selection=0x7ffc353015b0, cancelled=0x55d25ff1430c)
    at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-core/table-allgui.c:1269
#57 0x00007ffaf8d733f6 in gnucash_sheet_insert_cb
    (editable=0x55d25ff150b0, insert_text=0x55d2606998e0 "おやつ", insert_text_len=9, position=0x7ffc35301af4, sheet=0x55d25ff14260)
    at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-gnome/gnucash-sheet.c:1076
#58 0x00007ffaf7d5e146 in g_closure_invoke
    (closure=0x55d25f7fc0b0, return_value=0x0, n_param_values=4, param_values=0x7ffc353017a0, invocation_hint=0x7ffc35301720) at ../glib/gobject/gclosure.c:832
#59 0x00007ffaf7d8c2f8 in signal_emit_unlocked_R.isra.0
    (node=node@entry=0x55d25f707680, detail=detail@entry=0, instance=instance@entry=0x55d25ff150b0, emission_return=emission_return@entry=0x0, instance_and_params=instance_and_params@entry=0x7ffc353017a0) at ../glib/gobject/gsignal.c:3802
#60 0x00007ffaf7d7c095 in g_signal_emit_valist
    (instance=instance@entry=0x55d25ff150b0, signal_id=signal_id@entry=311, detail=<optimized out>, var_args=var_args@entry=0x7ffc353019a0)
    at ../glib/gobject/gsignal.c:3555
#61 0x00007ffaf7d7c544 in g_signal_emit_by_name (instance=instance@entry=0x55d25ff150b0, detailed_signal=detailed_signal@entry=0x7ffaf86295c7 "insert-text")
    at ../glib/gobject/gsignal.c:3654
#62 0x00007ffaf834969b in gtk_entry_insert_text (editable=0x55d25ff150b0, new_text=0x55d2606998c0 "おやつ", new_text_length=9, position=0x7ffc35301af4)
    at ../gtk/gtk/gtkentry.c:5175
#63 0x00007ffaf8350e28 in gtk_entry_enter_text (entry=0x55d25ff150b0, str=0x55d2606998c0 "おやつ") at ../gtk/gtk/gtkentry.c:6172
#64 0x00007ffaf7d5cb10 in g_cclosure_marshal_VOID__STRINGv
    (closure=0x55d25ff153a0, return_value=<optimized out>, instance=0x55d25ff14dc0, args=<optimized out>, marshal_data=<optimized out>, n_params=<optimized out>, param_types=0x55d25f717a80) at ../glib/gobject/gmarshal.c:1462
#65 0x00007ffaf7d7b451 in _g_closure_invoke_va
    (param_types=0x55d25f717a80, n_params=1, args=0x7ffc35301d30, instance=0x55d25ff14dc0, return_value=0x0, closure=0x55d25ff153a0)
    at ../glib/gobject/gclosure.c:895
#66 g_signal_emit_valist (instance=instance@entry=0x55d25ff14dc0, signal_id=signal_id@entry=331, detail=<optimized out>, var_args=var_args@entry=0x7ffc35301d30)
    at ../glib/gobject/gsignal.c:3462
#67 0x00007ffaf7d7c544 in g_signal_emit_by_name (instance=0x55d25ff14dc0, detailed_signal=0x7ffaf862a211 "commit") at ../glib/gobject/gsignal.c:3654
#68 0x00007ffaf7d5cb10 in g_cclosure_marshal_VOID__STRINGv
    (closure=0x55d25ffbfad0, return_value=<optimized out>, instance=0x55d25ff85cd0, args=<optimized out>, marshal_data=<optimized out>, n_params=<optimized out>, param_types=0x55d25f717a80) at ../glib/gobject/gmarshal.c:1462
#69 0x00007ffaf7d7b451 in _g_closure_invoke_va
    (param_types=0x55d25f717a80, n_params=1, args=0x7ffc35302040, instance=0x55d25ff85cd0, return_value=0x0, closure=0x55d25ffbfad0)
    at ../glib/gobject/gclosure.c:895
#70 g_signal_emit_valist (instance=0x55d25ff85cd0, signal_id=331, detail=<optimized out>, var_args=var_args@entry=0x7ffc35302040) at ../glib/gobject/gsignal.c:3462
#71 0x00007ffaf7d7c324 in g_signal_emit (instance=<optimized out>, signal_id=<optimized out>, detail=<optimized out>) at ../glib/gobject/gsignal.c:3612
#72 0x00007ffac2f8b58d in fcitx_im_context_commit_string (str=<optimized out>, context=0x55d25ff85cd0)
    at /usr/src/debug/fcitx5-gtk/fcitx5-gtk-5.0.23/gtk3/fcitximcontext.cpp:876
#73 fcitx_im_context_commit_preedit(FcitxIMContext*) (context=0x55d25ff85cd0) at /usr/src/debug/fcitx5-gtk/fcitx5-gtk-5.0.23/gtk3/fcitximcontext.cpp:890
#74 0x00007ffac2f8b6ab in fcitx_im_context_commit_preedit (context=0x55d25ff85cd0) at /usr/src/debug/fcitx5-gtk/fcitx5-gtk-5.0.23/gtk3/fcitximcontext.cpp:885
#75 fcitx_im_context_focus_out (context=0x55d25ff85cd0) at /usr/src/debug/fcitx5-gtk/fcitx5-gtk-5.0.23/gtk3/fcitximcontext.cpp:908
#76 fcitx_im_context_focus_out(GtkIMContext*) (context=0x55d25ff85cd0) at /usr/src/debug/fcitx5-gtk/fcitx5-gtk-5.0.23/gtk3/fcitximcontext.cpp:897
#77 0x00007ffaf83527f9 in gtk_entry_focus_out (widget=0x55d25ff150b0, event=event@entry=0x7ffc353024b0) at ../gtk/gtk/gtkentry.c:5015
#78 0x00007ffaf8287c89 in _gtk_marshal_BOOLEAN__BOXEDv
    (closure=0x55d25ee8c700, return_value=0x7ffc353022b0, instance=<optimized out>, args=<optimized out>, marshal_data=<optimized out>, n_params=<optimized out>, param_types=0x55d25ef1e620) at gtk/gtkmarshalers.c:130
#79 0x00007ffaf7d7b451 in _g_closure_invoke_va
    (param_types=0x55d25ef1e620, n_params=1, args=0x7ffc35302360, instance=0x55d25ff150b0, return_value=0x7ffc353022b0, closure=0x55d25ee8c700)
    at ../glib/gobject/gclosure.c:895
#80 g_signal_emit_valist (instance=0x55d25ff150b0, signal_id=77, detail=<optimized out>, var_args=var_args@entry=0x7ffc35302360) at ../glib/gobject/gsignal.c:3462
#81 0x00007ffaf7d7c324 in g_signal_emit (instance=instance@entry=0x55d25ff150b0, signal_id=<optimized out>, detail=detail@entry=0)
    at ../glib/gobject/gsignal.c:3612
#82 0x00007ffaf8544b25 in gtk_widget_event_internal.part.0.lto_priv.0 (widget=0x55d25ff150b0, event=0x7ffc353024b0) at ../gtk/gtk/gtkwidget.c:7812
#83 0x00007ffaf8d68712 in gnc_item_edit_focus_out (item_edit=0x55d25ff14990)
    at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-gnome/gnucash-item-edit.c:312
#84 0x00007ffaf8d73edc in gnucash_sheet_focus_out_event (widget=0x55d25ff14260, event=0x55d2606997f0)
    at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-gnome/gnucash-sheet.c:1275
#85 0x00007ffaf8284fd8 in _gtk_marshal_BOOLEAN__BOXED
    (closure=0x55d25ee8c700, return_value=0x7ffc35302670, n_param_values=<optimized out>, param_values=0x7ffc353026d0, invocation_hint=<optimized out>, marshal_data=<optimized out>) at gtk/gtkmarshalers.c:84
#86 0x00007ffaf7d5e146 in g_closure_invoke
    (closure=0x55d25ee8c700, return_value=0x7ffc35302670, n_param_values=2, param_values=0x7ffc353026d0, invocation_hint=0x7ffc35302650)
    at ../glib/gobject/gclosure.c:832
#87 0x00007ffaf7d8c427 in signal_emit_unlocked_R.isra.0
    (node=<optimized out>, detail=detail@entry=0, instance=instance@entry=0x55d25ff14260, emission_return=emission_return@entry=0x7ffc353027e0, instance_and_params=instance_and_params@entry=0x7ffc353026d0) at ../glib/gobject/gsignal.c:3841
#88 0x00007ffaf7d7baa0 in g_signal_emit_valist
    (instance=<optimized out>, signal_id=<optimized out>, detail=<optimized out>, var_args=var_args@entry=0x7ffc35302890) at ../glib/gobject/gsignal.c:3565
#89 0x00007ffaf7d7c324 in g_signal_emit (instance=instance@entry=0x55d25ff14260, signal_id=<optimized out>, detail=detail@entry=0)
    at ../glib/gobject/gsignal.c:3612
#90 0x00007ffaf8544b25 in gtk_widget_event_internal.part.0.lto_priv.0 (widget=0x55d25ff14260, event=0x55d2606997f0) at ../gtk/gtk/gtkwidget.c:7812
#91 0x00007ffaf8543d8b in gtk_widget_send_focus_change (widget=0x55d25ff14260, event=0x55d2606997f0) at ../gtk/gtk/gtkwidget.c:16244
#92 0x00007ffaf8552b52 in do_focus_change (widget=0x55d25ff14260, in=0) at ../gtk/gtk/gtkwindow.c:8487
#93 0x00007ffaf855320a in gtk_window_real_set_focus (window=<optimized out>, focus=0x55d25ffca9d0) at ../gtk/gtk/gtkwindow.c:8776
#94 0x00007ffaf7d5cca6 in g_cclosure_marshal_VOID__OBJECTv
    (closure=0x55d25ee6c510, return_value=<optimized out>, instance=0x55d25efe87b0, args=<optimized out>, marshal_data=<optimized out>, n_params=<optimized out>, param_types=0x55d25ef0bb20) at ../glib/gobject/gmarshal.c:1910
#95 0x00007ffaf7d7b451 in _g_closure_invoke_va
    (param_types=0x55d25ef0bb20, n_params=1, args=0x7ffc35302c40, instance=0x55d25efe87b0, return_value=0x0, closure=0x55d25ee6c510)
    at ../glib/gobject/gclosure.c:895
#96 g_signal_emit_valist (instance=0x55d25efe87b0, signal_id=114, detail=<optimized out>, var_args=var_args@entry=0x7ffc35302c40) at ../glib/gobject/gsignal.c:3462
#97 0x00007ffaf7d7c324 in g_signal_emit (instance=<optimized out>, signal_id=<optimized out>, detail=<optimized out>) at ../glib/gobject/gsignal.c:3612
#98 0x00007ffaf8517e6a in gtk_tree_view_grab_focus (widget=0x55d25ffca9d0) at ../gtk/gtk/gtktreeview.c:8823
#99 0x00007ffaf7d7b451 in _g_closure_invoke_va
    (param_types=0x0, n_params=0, args=0x7ffc35302ef0, instance=0x55d25ffca9d0, return_value=0x0, closure=0x55d25ee551a0) at ../glib/gobject/gclosure.c:895
#100 g_signal_emit_valist (instance=0x55d25ffca9d0, signal_id=57, detail=<optimized out>, var_args=var_args@entry=0x7ffc35302ef0) at ../glib/gobject/gsignal.c:3462
#101 0x00007ffaf7d7c324 in g_signal_emit (instance=instance@entry=0x55d25ffca9d0, signal_id=<optimized out>, detail=detail@entry=0)
    at ../glib/gobject/gsignal.c:3612
#102 0x00007ffaf853247c in gtk_widget_grab_focus (widget=0x55d25ffca9d0) at ../gtk/gtk/gtkwidget.c:8101
#103 0x00007ffaf8d63028 in popup_set_focus (widget=0x55d25ffc24c0, user_data=0x55d25fd89960)
    at /home/kazuki/ghq/github.com/Gnucash/gnucash/gnucash/register/register-gnome/combocell-gnome.c:1052
```

full version: [gnucash-1610.log.gz](https://github.com/Gnucash/gnucash/files/11315445/gnucash-1610.log.gz)

jralls commented:
I see the problem, but unfortunately I don't think the fix is as easy as this PR.

What's going on is that `gnc_combo_entry_type_ahead_search` calls `popup_set_focus` and setting focus on it. That removes focus from the GtkEntry, which emits a signal reporting that, and that's caught by the IME that commits its preedit causing the GtkEdit to update, sending an updated signal that `gnc_combo_entry_type_ahead_search` catches and it calls `popup_set_focus` again. It's not in the stack trace but somewhere in there focus is going back to the GtkEntry so each `popup_set_focus` makes it send its `unfocus` signal again. That might be a problem with fcitx5 not marking the preedit as committed or it might be in gnc-sheet.c's keypress code doing something dumb.

I'd think that the result of this PR would be that the popup wouldn't be able to get focus so you wouldn't be able to use the arrow keys to select a completion, you'd have to use the mouse.  Have you tested that?

kzkn commented:
> I'd think that the result of this PR would be that the popup wouldn't be able to get focus so you wouldn't be able to use the arrow keys to select a completion, you'd have to use the mouse. Have you tested that?

I tried it. Seems to be working correctly, what do you think? In the video, I am not touching the mouse.

![output](https://user-images.githubusercontent.com/2034773/234148936-8e412a3d-930a-4e6f-b9f7-c20525927c90.gif)

My code is the so-called Quick And Dirty approach and I don't think it is that so good. I don't think this PR needs to be merged if there is a better workaround.

jralls commented:
Let's try a more direct approach. Unfortunately the macOS IME doesn't cause the recursion so I can't replicate the issue. Can you try
```
--- a/gnucash/register/register-gnome/gnucash-item-edit.c
+++ b/gnucash/register/register-gnome/gnucash-item-edit.c
@@ -306,6 +306,9 @@ gnc_item_edit_focus_out (GncItemEdit *item_edit)
     g_return_if_fail (item_edit != NULL);
     g_return_if_fail (GNC_IS_ITEM_EDIT(item_edit));

+    if (item_edit->show_popup)
+        return; // Prevent recursion
+
     ev.type = GDK_FOCUS_CHANGE;
     ev.window = gtk_widget_get_window (GTK_WIDGET(item_edit->sheet));
     ev.in = FALSE;
```
in place of your change and see if it works?

kzkn committed:
Prevent infinite recursion from gnucash_item_edit_focus_out.

When I try to use IME (e.g. fcitx) to narrow down a combo box,
infinite recursion occurs via `popup_set_focus` and GnuCash falls down
in SIGSEGV.

ref: https://github.com/Gnucash/gnucash/pull/1610#issuecomment-1528080452

kzkn commented:
@jralls 
Great! It works fine :tada: 
I've pushed the code.

jralls commented:
Thanks!
